### PR TITLE
Support db_securityadmin

### PR DESCRIPTION
**Pull request by HarshLunagariya**
### Description

This commit adds support for db_securityadmin fixed database roles. Members of db_securityadmin can perform following operations within a particular database:
1. create schema
2. create database roles
3. alter/drop user-defined database roles
4. manage permissions via grant/revoke statements

Engine Changes - https://github.com/amazon-aurora/postgresql_modified_for_babelfish/pull/100

### Issues Resolved
BABEL-5135

### Test Scenarios Covered ###
* **Use case based -**

* **Boundary conditions -**

* **Arbitrary inputs -**

* **Negative test cases -**

* **Minor version upgrade tests -**

* **Major version upgrade tests -**

* **Performance tests -**

* **Tooling impact -**

* **Client tests -**

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 1]
```diff
@@ -5533,3 +5539,79 @@ pltsql_get_object_identity_event_trigger(ObjectAddress* address)
...
+		* Check if function's schema is from a different logical database and
+		* it is not a shared schema. If yes, then set userid to session user
+		* to allow cross database access.
+		*/
+		if (nspname == NULL ||
+			is_shared_schema(nspname) ||
+			!is_schema_from_db(schema_oid, get_cur_db_id()))
+		{
+			if (nspname)
+				pfree(nspname);
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/hooks.c]: pfree nspname in else block ?

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 2]
```diff
@@ -5533,3 +5539,79 @@ pltsql_get_object_identity_event_trigger(ObjectAddress* address)
...
+			return;
+		}
+		else
+		{
+			if (is_member_of_role(GetUserId(),
+								  get_role_oid(get_db_securityadmin_role_name(get_cur_db_name()), false)))
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/hooks.c]: get_curr_db_name() needs to be freed, should we use `get_current_pltsql_db_name`

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 3]
```diff
@@ -3003,13 +3003,16 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 					else if (isuser || isrole)
 					{
 						char *current_db_name = get_cur_db_name();
+						const char *db_securityadmin_role = get_db_securityadmin_role_name(current_db_name);
 
 						if (has_privs_of_role(GetUserId(), get_db_owner_oid(current_db_name, false)) ||
-						    (isuser && has_privs_of_role(GetUserId(), get_db_accessadmin_oid(current_db_name, false))))
+						    (isuser && has_privs_of_role(GetUserId(), get_db_accessadmin_oid(current_db_name, false))) ||
+							(isrole && has_privs_of_role(GetUserId(), get_role_oid(db_securityadmin_role, false))))
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: Should we create a helper to get db_security admin role OID ?

File: contrib/babelfishpg_tsql/src/dbcmds.c [hunk 4]
```diff
@@ -218,16 +230,19 @@ add_fixed_user_roles_to_bbf_authid_user_ext(const char *dbname)
...
 	dbo = get_dbo_role_name(dbname);
 	db_owner = get_db_owner_name(dbname);
 	db_accessadmin = get_db_accessadmin_role_name(dbname);
+	db_securityadmin = get_db_securityadmin_role_name(dbname);
 	guest = get_guest_role_name(dbname);
 
 	add_to_bbf_authid_user_ext(dbo, "dbo", dbname, "dbo", NULL, false, true, false);
 	add_to_bbf_authid_user_ext(db_owner, "db_owner", dbname, NULL, NULL, true, true, false);
 	add_to_bbf_authid_user_ext(db_accessadmin, DB_ACCESSADMIN, dbname, NULL, NULL, true, true, false);
+	add_to_bbf_authid_user_ext(db_securityadmin, DB_SECURITYADMIN, dbname, NULL, NULL, true, true, false);
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/dbcmds.c]: has_dbaccess should be false ? or does this role also grant connect privilege.

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 5]
```diff
@@ -5533,3 +5539,79 @@ pltsql_get_object_identity_event_trigger(ObjectAddress* address)
...
 }
+
+static void
+handle_grantstmt_for_dbsecadmin(ObjectType objType, Oid objId, Oid ownerId, AclMode privileges, Oid *grantorId, AclMode *grantOptions)
+{
+	ObjectAddress	address;
+	Oid				classid = InvalidOid;
+	Oid				schema_oid = InvalidOid;
+
+	if (!IS_TDS_CLIENT() || sql_dialect != SQL_DIALECT_TSQL)
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: IS_TDS_CLIENT is defined as 
```
#define IS_TDS_CLIENT() (*pltsql_protocol_plugin_ptr && \
						 (*pltsql_protocol_plugin_ptr)->is_tds_client)
```
we should use port->is_tds_conn now

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 6]
```diff
@@ -5533,3 +5539,79 @@ pltsql_get_object_identity_event_trigger(ObjectAddress* address)
...
+		return;
+
+	switch(objType)
+	{
+		case OBJECT_TABLE:
+		case OBJECT_SEQUENCE:
+		case OBJECT_COLUMN:
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: what about view or materialised view?

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 7]
```diff
@@ -5533,3 +5539,79 @@ pltsql_get_object_identity_event_trigger(ObjectAddress* address)
...
+			classid = RelationRelationId;
+			break;
+		case OBJECT_DOMAIN:
+		case OBJECT_TYPE:
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: Interesting! what kind of privilege could it have? and example/test case will help understand it better

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 8]
```diff
@@ -5533,3 +5539,79 @@ pltsql_get_object_identity_event_trigger(ObjectAddress* address)
...
+			classid = TypeRelationId;
+			break;
+		case OBJECT_FUNCTION:
+		case OBJECT_PROCEDURE:
+		case OBJECT_ROUTINE:
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/hooks.c]: whats difference from func/proc?

File: contrib/babelfishpg_tsql/src/multidb.c [hunk 9]
```diff
@@ -1451,6 +1451,21 @@ get_db_accessadmin_role_name(const char *dbname)
...
+{
+	char	   *name = palloc0(MAX_BBF_NAMEDATALEND);
+
+	if (get_migration_mode() == SINGLE_DB && strcmp(dbname, "master") != 0
+	    && strcmp(dbname, "tempdb") != 0 && strcmp(dbname, "msdb") != 0)
+		snprintf(name, MAX_BBF_NAMEDATALEND, "%s", DB_SECURITYADMIN);
+	else
+		snprintf(name, MAX_BBF_NAMEDATALEND, "%s_%s", dbname, DB_SECURITYADMIN);
+
+	truncate_identifier(name, strlen(name), false);
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/multidb.c:1512]: do we also truncate securityadmin for statement like grant/revoke and for procedure sp_addrolemember

Comment (Deepesh125) [reply contrib/babelfishpg_tsql/src/multidb.c:1512]: Please point me the test cases for this change

File: contrib/babelfishpg_tsql/src/dbcmds.c [hunk 10]
```diff
@@ -218,16 +230,19 @@ add_fixed_user_roles_to_bbf_authid_user_ext(const char *dbname)
 	const char     *dbo;
 	const char     *db_owner;
 	const char     *db_accessadmin;
+	const char     *db_securityadmin;
 	const char     *guest;
 
 	dbo = get_dbo_role_name(dbname);
 	db_owner = get_db_owner_name(dbname);
 	db_accessadmin = get_db_accessadmin_role_name(dbname);
+	db_securityadmin = get_db_securityadmin_role_name(dbname);
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/dbcmds.c:262]: pfree?

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 3 above)
Comment (Deepesh125) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: +1

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 1 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/hooks.c]: Addressed.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 2 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/hooks.c]: Addressed.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 6 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/hooks.c]: It's missing -- will add it, Just wondering how the view test case is working? Maybe it is getting the privileges via column privilege logic.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 5 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/hooks.c]: Done.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 7 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/hooks.c]: Currently, we don't support GRANT/REVOKE any type privileges. So we won't need it. We will only need if somewhere internally we grant some type permissions in case of tvp, table-valued functions etc. Need to add tests around that to confirm.

File: contrib/babelfishpg_tsql/src/dbcmds.c (hunk 10 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/dbcmds.c:262]: Done, in all relevant APIs.

File: contrib/babelfishpg_tsql/src/dbcmds.c (hunk 4 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/dbcmds.c]: No, it's doesn't grant connect privilege. Updated.

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 3 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: Done.

File: contrib/babelfishpg_tsql/src/multidb.c (hunk 9 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/multidb.c:1512]: Not covered in existing tests, Adding those.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 8 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/hooks.c]: [PG documentation](https://www.postgresql.org/docs/current/xproc.html#:~:text=Collectively%2C%20functions%20and%20procedures%20are%20also%20known%20as%20routines.%20There%20are%20commands%20such%20as%20ALTER%20ROUTINE%20and%20DROP%20ROUTINE%20that%20can%20operate%20on%20functions%20and%20procedures%20without%20having%20to%20know%20which%20kind%20it%20is.%20Note%2C%20however%2C%20that%20there%20is%20no%20CREATE%20ROUTINE%20command.)

> Collectively, functions and procedures are also known as routines. There are commands such as [ALTER ROUTINE](https://www.postgresql.org/docs/current/sql-alterroutine.html) and [DROP ROUTINE](https://www.postgresql.org/docs/current/sql-droproutine.html) that can operate on functions and procedures without having to know which kind it is. Note, however, that there is no CREATE ROUTINE command.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 5 above)
Comment (tanscorpio7) [reply contrib/babelfishpg_tsql/src/hooks.c]: Off topic, should we change `IS_TDS_CLIENT` to use port->is_tds_conn, or are they different.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 8 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/hooks.c]: Anyways, I'm trimming down on these object type which are not required.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 6 above)
Comment (tanscorpio7) [reply contrib/babelfishpg_tsql/src/hooks.c]: objType is set in parser and grant stmt uses the same syntax for views and tables, so views fall under table obj type.
I think we can match the list against [ExecGrantStmt_oids](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/4f6fa60f2f4465e0713f01667af667cd33bf8a7b/src/backend/catalog/aclchk.c#L593) to see if we covered everything.

File: contrib/babelfishpg_tsql/src/dbcmds.c [hunk 11]
```diff
@@ -346,13 +386,18 @@ gen_dropdb_subcmds(const char *dbname, List *db_users)
 	}
 
 	stmt = parsetree_nth_stmt(stmt_list, i++);
-	update_DropOwnedStmt(stmt, list_make3(pstrdup(db_accessadmin), pstrdup(db_owner), pstrdup(dbo)));
+	update_DropOwnedStmt(stmt, list_make4(pstrdup(db_securityadmin), pstrdup(db_accessadmin), pstrdup(db_owner), pstrdup(dbo)));
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/dbcmds.c:391]: Why isn't datareader and writer not here ?

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 12]
```diff
@@ -176,6 +176,7 @@ static bool pltsql_bbfCustomProcessUtility(ParseState *pstate,
 									  ParamListInfo params, QueryCompletion *qc);
 extern void pltsql_bbfSelectIntoUtility(ParseState *pstate, PlannedStmt *pstmt, const char *queryString, 
 					QueryEnvironment *queryEnv, ParamListInfo params, QueryCompletion *qc, ObjectAddress *address);
+static void handle_grantstmt_for_dbsecadmin(ObjectType objType, Oid objId, Oid ownerId, AclMode privileges, Oid *grantorId, AclMode *grantOptions);
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/hooks.c:180]: Would this be more suited to declare under Object Access Hook comment

Also lets add some comments in utility hook under GRANTSTMT that indicates there is some lower level handling for members of dbsecadmin for this command type.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 6 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/hooks.c]: @tanscorpio7 Thanks for pointing it. Unknowingly, I have done that only. But I think there is not harm including those extra object types.

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 13]
```diff
@@ -5533,3 +5539,81 @@ pltsql_get_object_identity_event_trigger(ObjectAddress* address)
...
+			!is_schema_from_db(schema_oid, get_cur_db_id()))
+		{
+			if (nspname)
+				pfree(nspname);
+			return;
+		}
+		else
+		{
+			if (is_member_of_role(GetUserId(),
+								  get_role_oid(get_db_securityadmin_role_name(get_current_pltsql_db_name()), false)))
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/hooks.c]: Would it be better to add these checks at beginning of the function ?
quick exit 
-> if dialect not tsql 
-> if already have the privs
-> if not member of dbsecadmin
So that this code gets triggered as less as possible.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 6 above)
Comment (tanscorpio7) [reply contrib/babelfishpg_tsql/src/hooks.c]: I think it may be better to not add those since we cannot add tests for it.

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 13 above)
Comment (tanscorpio7) [reply contrib/babelfishpg_tsql/src/hooks.c]: Also lets add some comments for this handle_grantstmt_for_dbsecadmin

File: contrib/babelfishpg_tsql/src/dbcmds.c (hunk 11 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/dbcmds.c:391]: Adding @shalinilohia50 here for better context

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 13 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/hooks.c]: Done

File: test/JDBC/expected/db_securityadmin-vu-verify.out [hunk 14]
```diff
@@ -0,0 +1,851 @@
...
+
+DROP USER babel_5135_u1;
+GO
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: Cannot drop the user 'babel_5135_u1', because it does not exist or you do not have permission.)~~
+
+
+-- CASE 4 - CREATE SCHEMA should be allowed
+CREATE SCHEMA babel_5135_sch1;
```
Comment (shalinilohia50) [inline test/JDBC/expected/db_securityadmin-vu-verify.out:222]: 1. Who will be the owner of schema in this case? dbo or the current user of the login? Can we add tests for the same?
2. If there is no mapped user for a login, can we add guest as a member of db_securityadmin? If yes, will guest be the owner of the schema?

File: contrib/babelfishpg_tsql/src/dbcmds.c [hunk 15]
```diff
@@ -295,20 +327,26 @@ gen_dropdb_subcmds(const char *dbname, List *db_users)
 		char	   *user_name = (char *) lfirst(elem);
 
 		if (strcmp(user_name, db_owner) != 0 && strcmp(user_name, dbo) != 0 &&
-		    strcmp(user_name, db_accessadmin) != 0 && strcmp(user_name, db_datareader) != 0 &&
+			strcmp(user_name, db_accessadmin) != 0 &&
+			strcmp(user_name, db_securityadmin) != 0 &&
+		    strcmp(user_name, db_datareader) != 0 &&
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/dbcmds.c]: Nit: Indentation.

File: contrib/babelfishpg_tsql/src/dbcmds.c [hunk 16]
```diff
@@ -295,20 +327,26 @@ gen_dropdb_subcmds(const char *dbname, List *db_users)
...
 			strcmp(user_name, db_datawriter) != 0)
 		{
 			appendStringInfo(&query, "DROP OWNED BY dummy CASCADE; ");
 			appendStringInfo(&query, "DROP ROLE dummy; ");
 			expected_stmts += 2;
 		}
 	}
-	appendStringInfo(&query, "DROP OWNED BY dummy, dummy, dummy CASCADE; ");
+	appendStringInfo(&query, "DROP OWNED BY dummy, dummy, dummy, dummy CASCADE; ");
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/dbcmds.c:341]: Do we really need this `DROP OWNED BY db_securityadmin`?
IIUC, the dependent objects should be correctly cleaned up by executing `DROP OWNED BY dbo, dbo_owner`.

File: contrib/babelfishpg_tsql/src/dbcmds.c [hunk 17]
```diff
@@ -334,7 +372,9 @@ gen_dropdb_subcmds(const char *dbname, List *db_users)
 		char	   *user_name = (char *) lfirst(elem);
 
 		if (strcmp(user_name, db_owner) != 0 && strcmp(user_name, dbo) != 0 &&
-		    strcmp(user_name, db_accessadmin) != 0 && strcmp(user_name, db_datareader) != 0 &&
+			strcmp(user_name, db_accessadmin) != 0 &&
+			strcmp(user_name, db_securityadmin) != 0 &&
+		    strcmp(user_name, db_datareader) != 0 &&
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/dbcmds.c]: Nit: Indentation

File: contrib/babelfishpg_tsql/src/dbcmds.c (hunk 11 above)
Comment (shalinilohia50) [reply contrib/babelfishpg_tsql/src/dbcmds.c:391]: Do we really need this `DROP OWNED BY db_securityadmin/ db_accessadmin`?

The dependent objects should be correctly cleaned up by executing `DROP OWNED BY dbo, dbo_owner`.
There was no need to add `db_datareader` and `db_datawriter` here because of the fact that the objects like role/user that could have been dependent on these database roles were getting dropped already using the above statement.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 5 above)
Comment (Deepesh125) [reply contrib/babelfishpg_tsql/src/hooks.c]: Yes, we generally should fix that but not with this PR. @tanscorpio7 could you please open jira for that? we need to evaluate why is current check making use of (*pltsql_protocol_plugin_ptr)->is_tds_client and what challenges we may face if we use tds_conn?

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 6 above)
Comment (Deepesh125) [reply contrib/babelfishpg_tsql/src/hooks.c]: +1

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 18]
```diff
@@ -3306,13 +3308,13 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 						char		*db_name = get_cur_db_name();
 						bool 		is_member_of_db_owner = false;
 						bool 		is_member_of_db_accessadmin = false;
+						bool 		is_member_of_db_securityadmin = false;
 						int 		save_sec_context;
 						Oid 		save_userid;
 						Oid 		db_owner = get_db_owner_oid(db_name, false);
 						Oid 		db_accessadmin = get_db_accessadmin_oid(db_name, false);
+						Oid			db_securityadmin = get_role_oid(get_db_securityadmin_role_name(db_name), false);
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: `get_db_securityadmin_oid` here and some other places.

File: contrib/babelfishpg_tsql/src/multidb.c (hunk 9 above)
Comment (Deepesh125) [reply contrib/babelfishpg_tsql/src/multidb.c:1512]: Apart from that, I think we can save some calls to truncate_identifier if we can cache db_securityadmin rolname for given db

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 19]
```diff
@@ -5533,3 +5540,96 @@ pltsql_get_object_identity_event_trigger(ObjectAddress* address)
...
+	if (OidIsValid(schema_oid))
+	{
+		char *nspname = get_namespace_name(schema_oid);
+		/*
+		 * Don't allow if object's schema is not from current database OR
+		 * it is a shared schema.
+		 */
+		if (nspname == NULL ||
+			is_shared_schema(nspname) ||
+			!is_schema_from_db(schema_oid, get_cur_db_id()))
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/hooks.c]: Doesnt `is_schema_from_db(schema_oid, get_cur_db_id()`  condition make the other two redundant ?

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/dbcmds.c (hunk 11 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/dbcmds.c:391]: Database cleanup should not be blocked because of these internal roles's object dependency. So, we have added this for sanity check.

File: contrib/babelfishpg_tsql/src/dbcmds.c (hunk 16 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/dbcmds.c:341]: It just be on the safer side and not to block the database cleanup due to it.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 19 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/hooks.c]: Addressed.

File: contrib/babelfishpg_tsql/src/dbcmds.c (hunk 15 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/dbcmds.c]: Addressed.

File: contrib/babelfishpg_tsql/src/dbcmds.c (hunk 17 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/dbcmds.c]: Addressed.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 12 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/hooks.c:180]: Would this be more suited to declare under Object Access Hook comment --> No, I think it is better placed under utility only.

Added comments in GRANT stmt handling in bbf_Processutility code.

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 18 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: Done

File: test/JDBC/expected/db_securityadmin-vu-verify.out (hunk 14 above)
Comment (HarshLunagariya) [reply test/JDBC/expected/db_securityadmin-vu-verify.out:222]: 1. current_user, Added the query where we are validating the schema owner.
2. Yes for both sub-questions. Added the test case for the same.

File: contrib/babelfishpg_tsql/src/dbcmds.c [hunk 20]
```diff
@@ -366,6 +413,13 @@ gen_dropdb_subcmds(const char *dbname, List *db_users)
 	stmt = parsetree_nth_stmt(stmt_list, i++);
 	update_DropRoleStmt(stmt, dbo);
 
+	pfree((char *) dbo);
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/dbcmds.c:416]: Should we define all these variables as (char *) instead of (const char *)?

File: contrib/babelfishpg_tsql/src/dbcmds.c [hunk 21]
```diff
@@ -1417,12 +1471,14 @@ create_db_roles_in_database(const char *dbname, List *parsetree_list)
 	int            i = 0;
 	char           *db_owner;
 	char           *db_accessadmin;
+	char		   *db_securityadmin;
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/dbcmds.c:1474]: Nit: Indentation

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 22]
```diff
@@ -5533,3 +5540,89 @@ pltsql_get_object_identity_event_trigger(ObjectAddress* address)
...
+		schema_oid = get_object_namespace(&address);
+	}
+
+	if (OidIsValid(schema_oid))
+	{
+		/*
+		 * Don't allow if object's schema is not from current database OR
+		 * it is a shared schema.
+		 */
+		if (!is_schema_from_db(schema_oid, get_cur_db_id()))
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/hooks.c:5607]: Does this check handle shared schema as well?

File: contrib/babelfishpg_tsql/src/dbcmds.c (hunk 20 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/dbcmds.c:416]: Keeping it same for now since it's an improvement but not affect the operations.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 22 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/hooks.c:5607]: It will return false.
